Fix hardcoded sha1sum command and compare sha1 value

ChinYikMing opened the pull request.

ChinYikMing committed:
Fix hardcoded sha1sum command and compare sha1 value

jserv commented on mk/external.mk:
```diff
@@ -23,18 +23,26 @@ QUAKE_DATA_VERIFY = echo "$(strip $$($(T)_DATA_SHA1))  $$@" | $(SHA1SUM) -c
-TIMIDITY_DATA_VERIFY = echo "$(TIMIDITY_DATA_SHA1)" > $(TIMIDITY_TMP_FILE) | find $(TIMIDITY_DATA) -type f -print0 | sort -z | xargs -0 shasum | shasum | cut -f 1 -d ' '
+TIMIDITY_TMP_FILE1 = /tmp/timidity_sha1.txt
+TIMIDITY_TMP_FILE2 = /tmp/timidity_sha2.txt
+TIMIDITY_DATA_VERIFY = echo "$(TIMIDITY_DATA_SHA1)" > $(TIMIDITY_TMP_FILE1) \
```
Can you make it modular? That is, provide calls for zip and (tar.)gz respectively?

jserv commented on mk/external.mk:
```diff
@@ -1,41 +1,113 @@
+
+# temporarily files to store correct SHA1 value and computed SHA1 value
+# respectively for verification of directory source
+SHA1_FILE1 := /tmp/sha1_file1
```
Use `mktemp` instead.

jserv commented on mk/external.mk:
```diff
@@ -1,41 +1,110 @@
+define extract
+$(eval COMPRESSED_SUFFIX := $(suffix $(1)))
+$(eval COMPRESSED_IS_ZIP := $(filter $(COMPRESSED_SUFFIX),.zip))
+$(eval _ := $(if $(COMPRESSED_IS_ZIP),                  \
```
Indent with spaces.

jserv commented on mk/external.mk:
```diff
@@ -1,41 +1,110 @@
+define verify
+    $(eval COMPRESSED_IS_DIR := $(if $(wildcard $(2)/*),1,0))
+    $(eval _ := $(if $(filter 1,$(COMPRESSED_IS_DIR)),             \
+		(                                                          \
```
Can you reduce the number of spaces? That is, I would expect more compact indetions.

ChinYikMing replied:
> Can you reduce the number of spaces? That is, I would expect more compact indetions.

Do you mean to not align the backslash?

jserv replied:
> Do you mean to not align the backslash?

No, I meant something like this:
```
define extract
    $(eval COMPRESSED_SUFFIX := $(suffix $(1)))
    $(eval COMPRESSED_IS_ZIP := $(filter $(COMPRESSED_SUFFIX),.zip))
    $(eval _ := \
        $(if $(COMPRESSED_IS_ZIP), \
            ($(eval EXTRACTOR := unzip -d $(OUT) $(1))), \
            ($(eval EXTRACTOR := tar -xf $(1) -C $(OUT))) \
    ))
    $(eval _ := $(shell $(EXTRACTOR)))
endef
```

ChinYikMing replied:
> > Do you mean to not align the backslash?
> 
> No, I meant something like this:
> 
> ```
> define extract
>     $(eval COMPRESSED_SUFFIX := $(suffix $(1)))
>     $(eval COMPRESSED_IS_ZIP := $(filter $(COMPRESSED_SUFFIX),.zip))
>     $(eval _ := \
>         $(if $(COMPRESSED_IS_ZIP), \
>             ($(eval EXTRACTOR := unzip -d $(OUT) $(1))), \
>             ($(eval EXTRACTOR := tar -xf $(1) -C $(OUT))) \
>     ))
>     $(eval _ := $(shell $(EXTRACTOR)))
> endef
> ```

Got it.

jserv commented on mk/external.mk:
```diff
@@ -1,41 +1,98 @@
+    $(eval COMPRESSED_SUFFIX := $(suffix $(1)))
+    $(eval COMPRESSED_IS_ZIP := $(filter $(COMPRESSED_SUFFIX),.zip))
+    $(eval _ :=  \
+		$(if $(COMPRESSED_IS_ZIP), \
```
Use 4 spaces for indention.

ChinYikMing committed:
Modularize the extractor and verifier for external

Abstract the extractor and verifier to support .zip and .gz compressed
files.

The $(call) function executes during the expansion, so we have to make
all steps inside download-extract-verify to $(call) like, so that they
can be run in sequentially. To do that, introducing three more functions
which are prologue, download, and epilogue.

jserv commented:
Thank @ChinYikMing for contributing!
